Fix "serverless function has crashed" error

goranstevanovic opened the pull request:
I think I've managed to (partially) fix the issue where the app doesn't appear to start and we only see `This Serverless Function has crashed` error message.

There were a few things I had to change. I created a new shadow database in the Supabase using `psql` command. I've added a new environment variable (`SHADOW_DATABASE_URL`) that points to that new database. I've changed the Prisma schema file to include shadow database URL. I've also added RHEL as a binary target, because there was an error message indicating that we had to add this.

This was not enough, as there was now a new error message, something about not been able to convert string to int, I think. So, I changed the `User` model's `id` field definition from String type that uses `uuid()` to get the value to Int type that uses `autoincrement()` to get the value. That appears to fix the issue.

I was able to create user, to list all users, and to delete all users when running the app locally.

Using the deployed app on Vercel is another story. It looks like there are always two deployments made on Vercel whenever we push changes to GitHub. The first one happens because we connected the Vercel project with the GitHub repository via Vercel project settings. The other one happens because of the GitHub actions.

The first one shows commit message in the third column in the [Project Deployments page](https://vercel.com/syntaxerrorba-protonmailc/kid-toys-service/deployments). The second one always show `via Vercel CLI`.

The app works when checking the second kind of deployments (`via Vercel CLI`). I still get a `serverless function has crashed` error message when checking the first kind of deployments. I guess this means that our GitHub actions are OK.

I'm not sure what's wrong with the first type of deployments. I thought maybe the application doesn't get build correctly, so I added a build script, but that didn't help. I used `tsc` as a build command, but I think we might have to use `vercel build` for build script. GitHub action uses that command and it works.

goranstevanovic committed:
Add RHEL as binary target to Prisma Schema file

goranstevanovic committed:
Create user in 'user' table on application start

goranstevanovic committed:
Update User model's id field definition

goranstevanovic committed:
Revert User model's id field definition

goranstevanovic committed:
Change User model's id field type from String to Int

goranstevanovic committed:
Add shadow database URL to Prisma Schema

goranstevanovic committed:
Create new Prisma migration from scratch

goranstevanovic committed:
Get users from User table on application start

goranstevanovic committed:
Add npm run build script

goranstevanovic committed:
Change main file location

GoodbyePlanet commented:
I have a question, why do you need a shadow database URL?

```It looks like there are always two deployments made on Vercel whenever we push changes to GitHub.```
This is because of 
```"git": {
    "deploymentEnabled": {
      "main": false
    }
```
on your branch, vercel doesn't know that he should avoid automatic deployment when you push to yuor branch

goranstevanovic commented:
> I have a question, why do you need a shadow database URL?

I followed the [Supabase's Prisma guide](https://supabase.com/docs/guides/integrations/prisma):
> By default, Prisma migrations will try to drop the postgres database, which can lead to conflicts with Supabase databases. For this scenario, use [Prisma Shadow Databases](https://www.prisma.io/docs/concepts/components/prisma-migrate/shadow-database#cloud-hosted-shadow-databases-must-be-created-manually)

goranstevanovic committed:
Disable Prisma commands in index.ts

goranstevanovic commented:
I commented out commands that interact with the database, and now both deployments work. :confused:

goranstevanovic committed:
Uncomment commands to get all users from database in index.ts

goranstevanovic committed:
Wrap user fetching in try/catch block

goranstevanovic committed:
Add /add endpoint for creating users

goranstevanovic commented:
I've created `/add` endpoint that will always create the same user, just for testing purposes.

goranstevanovic commented:
Deployment `via Vercel CLI` still works.

Non-Vercel CLI deployment still doesn't work. It does work if there are no users in the database. If the database is not empty, it does not work. It shows an error `Error converting field "id" of expected non-nullable type "String", found incompatible value of "XX".` (XX being some number). I'm not sure why it's expecting a value of type String, when it's set to Int in Prisma schema file. [Prisma documentation](https://www.prisma.io/typescript) says that the correct TypeScript type will be created based on the model in the schema file.

goranstevanovic committed:
Change destination for all routes in Vercel settings

goranstevanovic committed:
Update build settings in Vercel config

goranstevanovic committed:
Update routes dest field in Vercel config

goranstevanovic committed:
Restore builds use and routes dest fields to previous values

goranstevanovic committed:
Change Vercel settings

goranstevanovic committed:
Update Vercel settings

goranstevanovic committed:
Fix typo and rename all error variables to error in index.ts

goranstevanovic committed:
Update Vercel settings

goranstevanovic commented:
Just an update on things I have done. I've changed the `builds` and `routes` sections in Vercel settings.

`builds` was pointing to `index.js`, not `index.ts`, which we are using. `routes.dest` property also didn't point to correct destination (compiled) file. I also updated `Build Command` and `Output Directory` in [general project settings on Vercel](https://vercel.com/syntaxerrorba-protonmailc/kid-toys-service/settings/general). I had to define `build` script in `package.json` to use TypeScript compiler to compile our code. Also, the default output directory on Vercel project settings is not the same as the one we defined in TypeScript settings.

Now the main page works in both type of deployments, but accessing any other route except for the main `/` route doesn't work. I'm not sure how to set up the `routes` settings so it works, but [Vercel documentations shows that `routes` property is now legacy](https://vercel.com/docs/project-configuration#legacy/routes), so we have to change it to something else.

goranstevanovic committed:
Remove routes section from Vercel settings

goranstevanovic committed:
Add rewrites section to Vercel settings

goranstevanovic committed:
Fix rewrites section in Vercel settings

goranstevanovic committed:
Update rewrites section in Vercel settings

goranstevanovic commented:
I've changed the `rewrites.destinaton` field in Vercel settings back to `index.ts` from `dist/index.js` and now seems that everything works.

goranstevanovic committed:
Update server listening message

goranstevanovic commented:
> I also updated Build Command and Output Directory in [general project settings on Vercel](https://vercel.com/syntaxerrorba-protonmailc/kid-toys-service/settings/general).

I've changed this back to default settings. It appears that that wasn't a cause of these issues.

GoodbyePlanet commented:
> I've changed the `rewrites.destinaton` field in Vercel settings back to `index.ts` from `dist/index.js` and now seems that everything works.

Nice job :) I will approve PR and you can merge it to the main branch so that you can continuer working on other tickets.

GoodbyePlanet commented on index.ts:
```diff
@@ -9,30 +9,39 @@ async function connectDb() {
+  }
+});
+
+app.get("/add", async (req, res) => {
```
Since this is just a test endpoint can you delete it, and also /delete endpoint, in the tickets in TODO section you will have a chance to implement new endpoints :)

goranstevanovic replied:
I removed those two endpoints.

goranstevanovic committed:
Remove /add and /delete endpoints

goranstevanovic commented:
@GoodbyePlanet, since this PR request contains a lot of commits, should I choose 'Squash and merge' option when merging, so we only have one commit in history, instead of 20+?

GoodbyePlanet commented:
Yup, just squash and merge it.
On Tue, 17 Jan 2023 at 19:22, Goran Stevanovic ***@***.***>
wrote:

> @GoodbyePlanet <https://github.com/GoodbyePlanet>, since this PR request
> contains a lot of commits, should I choose 'Squash and merge' option when
> merging, so we only have one commit in history, instead of 20+?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/z0bla/kid-toys-service/pull/36#issuecomment-1385844181>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFP6XYF73YSGGV6OB46QKE3WS3PPJANCNFSM6AAAAAAT35XLNA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

goranstevanovic commented:
So, I merged this branch with `development` branch. Deployment `via Vercel CLI` works. The other type of deployment doesn't work. :smirk: There's that error message `Error converting field "id" of expected non-nullable type "String", found incompatible value of "XX"` again. :roll_eyes: 

I'm out of ideas what could be wrong. :confused:
